### ISSUE TITLE
Suggestion: si pas de légende dans une image, indiquer le nom du fichier dans la légende

### ISSUE THREAD

**Issue by jpclerc39** [l-atelier-des-chercheurs/dodoc#457]
dans le but d'avoir une info sur le fichier média sans avoir à l'ouvrir, dans le cas où on a pas d'aperçu ou même comme information complémentaire facile à voir, cela pourrait être activable ou non dans les options générales.

**Comment by louis-ev** [comment]
J’ai peur que ça ajoute beaucoup de bruit pour peu d’information (les noms des médias qui viennent de la page capture n’ont pas d’intérêt, seuls ceux des fichiers importés potentiellement).

Les fichiers sans aperçus affichent déjà leur nom à la place de la vignette non ?

<img width="351" alt="image" src="https://user-images.githubusercontent.com/1948417/73262360-41216f00-41ce-11ea-91a0-0cb2cef4a1d6.png">

Tu peux donner un cas précis où ça rendrait service ?

**Comment by jpclerc39** [comment]
je parlais en fait des fichiers importés, mais en y réfléchissant, non on va laisser cette demande de coté si tu veux bien ou la fermer, le ou les cas rencontrés ne sont pas assez représentatifs

**Comment by louis-ev** [comment]
Ça marche.
J’en profite pour te demander : un aperçu des STL aurait-il du sens ? Je peux fouiller pour voir s’il existe des paquets qui peuvent produire une image à partir d’un STL pour donner une idée du contenu. Utile pour les DXF aussi ? Si tu pouvais faire une liste au pere noel elle contiendrait quoi :D

**Comment by jpclerc39** [comment]
ah mais ouiiii ça m’intéresse pour le fablab, avec les fichiers qu'on trouve chez les amis makers, alors rapidement:
STL
DXF 
SVG

je pense à d'autres fichiers mais plus propriétaires et liés à des applications spécifiques

**Comment by louis-ev** [comment]
Pour les STL, la bibliothèque suivante a l’air de plutôt bien marcher : https://hujiulong.github.io/vue-3d-model/#/demo-basic (sous licence MIT donc intégrable à do•doc). Après, ce plugin permet un aperçu "3d" dynamique, donc ça ne créé pas un aperçu jpeg — dans la bibliothèque il y aurai pas d’aperçu.
Ce plugin là si : https://github.com/fakhrullah/node-stl-to-thumbnail

Pour les DXF : https://github.com/gdsestimating/three-dxf (y a pas de démo en ligne). Licence MIT aussi. Pareil pour l’aperçu, il faudrait voir côté serveur si on peut récupérer l’image et la stocker.

SVG : ça c’est bon normalement :) Y a l’aperçu dans la bibliothèque maintenant, par contre pas de player pour naviguer dedans, zoomer, etc. C’est important à ton avis ?

**Comment by jpclerc39** [comment]
si intégration des STL, la vue dynamique me paraît indispensable, à prioriser par rapport à un aperçu vignette/thumbnails si les deux ne sont pas possibles (et là j'imagine une fonctionnalité qui permettrait de faire un snapshot de l'objet visualisé et s'en servir comme image bibliothèque)

amha, pour le SVG non pas besoin de se promener dedans, l'aperçu est suffisant 

et en idée d'intégration future, ce serait intéressant d'intégrer un petit visualiseur de texte/code (je pense aux fichiers type .ino -Ide Arduino, .Gcode  (ces fichiers sont très orientés fablab/prototypage....)

**Comment by jubonhomme** [comment]
+1 pour un aperçu de SVG / DXF / STL
Dans le cadre du projet Fablab à l'école par exemple, les classes documentent les réalisations avec des images / vidéos et publications. Il y a déjà quelques fichiers SVG ou DXF dans les différents dossiers des projets pour que les utilisateurs puissent télécharger directement le fichier à modifier / utiliser pour la découpe ou l'impression... façon plateforme Thingiverse
Pouvoir visualiser les fichier SVG / DXF / STL directement dans Dodoc serait vraiment très pratique.
L'indication du poids du fichier serait aussi intéressante avant de télécharger un fichier (mais je crois avoir vu passer la demande dans un autre message)

@louis-ev 
<lettreauPèreNoël> Si tu ajoutes des bibliothèques pour avoir des aperçus, est-ce que ça rendra possible l'intégration de ces médias dans une publication, par exemple dans un encart comme un [3D pdf](https://youtu.be/50b4IMzbWE0?t=76) simplifié ?  </lettreauPèreNoël>

+1 aussi pour un aperçu des fichiers txt

**Comment by louis-ev** [comment]
Ha ben tiens alors, je connaissais pas les 3D pdf !
Aucune chance que ça marche non, les bibliothèques qu’on utilise gèrent pas ça…
Au mieux il y aura un aperçu jpeg ou png du contenu, et ça sera difficile de faire mieux.

**Comment by louis-ev** [comment]
Par contre c’est vrai que dans l’export "page web" on aura peut-être envie d’avoir le lecteur en 3d et pas juste un aperçu. À voir ce qui est possible.

**Comment by louis-ev** [comment]
Hello !

Par un samedi après-midi ensoleillé de confinement, voici un premier aperçu de cette fonctionnalité : 

- en petit
<img width="168" alt="image" src="https://user-images.githubusercontent.com/1948417/78454705-05a59400-769a-11ea-8f0d-e60164979d21.png">

- en grand
<img width="1705" alt="image" src="https://user-images.githubusercontent.com/1948417/78454710-0d653880-769a-11ea-8574-b648b292f26d.png">

Pas de vue dynamique encore, mais ça devrait être + facile à ajouter. Le truc est de regarder le poids supplémentaire dans le fichier JS envoyé au client — celui qui permet de construire l’interface de do•doc. Il pèse 600ko compressé, c’est pas négligeable, et si ça représente 200 ou 300 kb de plus faudra trouver une autre approche.

**Comment by jpclerc39** [comment]
yes ! merci pour cette évolution & fonctionnalité, je reviens un petit coup par ici (très pris par la fabrication de visières via le fablab ces derniers temps...)
tu nous diras si c'est dispo sur le serveur test que j'aille faire joujou avec des fichiers stl

**Comment by louis-ev** [comment]
En cours de finalisation : 

<img width="1408" alt="image" src="https://user-images.githubusercontent.com/1948417/78473256-6aff9080-773f-11ea-8a41-57c481b7f7b6.png">

Je te dis quand c’est déployé.

**Comment by louis-ev** [comment]
Et voilà : 
https://test.dodoc.fr/test-stl
(c’est toi qui a créé ce projet je suppose ?)

**Comment by louis-ev** [comment]
OBJ ça paraît toujours utile ?

**Comment by jubonhomme** [comment]
Fabuleux le STL !

**Comment by jubonhomme** [comment]
C'est top même sur tablette android en tactile avec un et deux doigts pour les déplacements... vraiment très réussi !

**Comment by jpclerc39** [comment]
> Et voilà :
> https://test.dodoc.fr/test-stl
> (c’est toi qui a créé ce projet je suppose ?)

oui :-) 
c'est excellent bravo, euh tu sors quand la prochaine release ^_^
bon j'en remonte un peu, des gros pour voir, j'espère ne pas trop surcharger, ce sont des fichiers en lien avec les projets impression #3D solidaires du moment

**Comment by louis-ev** [comment]
Chouette :)
l’aperçu "image" et le mode interactif ont un rendu un peu différent, je dois pouvoir ajuster le rendu pour qu’ils soit plus proche (l’image + ressemblante au interactif)

**Comment by jubonhomme** [comment]
Perso je trouve ça très bien que l’aperçu image (au trait) soit différent de celui interactif (en plein).
Le dessin au trait est plus lisible quand on ne peut pas le bouger.
Avoir une différence permet de mieux se repérer visuellement entre les deux.

**Comment by jubonhomme** [comment]
Les fichiers déposés avant ne s'affichent pas, ils faut les réimporter pour que ce soit pris en compte.

**Comment by jpclerc39** [comment]
bon j'ai chargé la mule, avec des fichiers de 10m, ça marche bien :-) un peu long forcément, attention il manque une option: "Envoyer vers l'imprimante #3D" ;-)

**Comment by jubonhomme** [comment]
Il suffit de connecter l'imprimante au dossier dodoc ;)

**Comment by jubonhomme** [comment]
Bravo pour les embouts à masque de plongée décathlon

**Comment by louis-ev** [comment]
> Les fichiers déposés avant ne s'affichent pas, ils faut les réimporter pour que ce soit pris en compte.
> 
> 
> 
> Exemple : https://test.dodoc.fr/test-dodoc-9-1/media/truck-ecrou-standard*stl*txt

Oui, malheureusement ils ont pour type « autre ». J’ai ajouté un nouveau type, le STL, donc il faut soit changer manuellement leur type en modifiant le fichier meta.txt, soit les réimporter.

Avez vous besoin du support des OBJ ? Je comptais regarder pour les DXF ensuite.

**Comment by jubonhomme** [comment]
Les seuls formats que j'utilise pour l'instant sont le SVG, le STL et le DXF, essentiellement pour de l'impression couleur, de l'impression 3D et de la découpe laser.
Je ne connais pas le .OBJ, c'est de la 3D avec texture c'est ça ?
Vous en faites quelle utilisation ?

**Comment by louis-ev** [comment]
Ha mais je me souvenais pas de cette demande pour les fichiers ino, gcode et compagnie ! J’y réfléchis.

**Comment by jubonhomme** [comment]
Oui pour un éditeur de code assez light...
Exemple de celui-ci :
 https://www.w3schools.com/css/tryit.asp?filename=trycss3_border-radius

**Comment by jubonhomme** [comment]
Concernant les aperçus des fichiers .STL :
Lorsque l'on voit l'image on voit le bouton en bas aperçu interactif et inversement.
Perso ça me fais bizarre de le voir dans ce sens, on a l'impression que c'est une légende et que l'on est sur l'autre aperçu.
Peut-on imaginer à la place un bouton double interactif / image avec un côté grisé et un côté coloré en fonction de celui sélectionné... comme ça on voit les deux affichages possibles et celui actif qui correspond à la vue réelle.

**Comment by jpclerc39** [comment]
> Ha mais je me souvenais pas de cette demande pour les fichiers ino, gcode et compagnie ! Qui m’arrangerai bien aussi d’ailleurs.
> J’y réfléchis, techniquement je sais pas bien comment m’y prendre. L’idéal ne serait-il pas un petit éditeur de code qui permette de modifier le contenu de ces fichiers là ?

> Les seuls formats que j'utilise pour l'instant sont le SVG, le STL et le DXF, essentiellement pour de l'impression couleur, de l'impression 3D et de la découpe laser.
> Je ne connais pas le .OBJ, c'est de la 3D avec texture c'est ça ?
> Vous en faites quelle utilisation ?
> 
> Donc pour moi le DXF serait le bienvenue, ainsi qu'un petit lecteur/éditeur de .TXT comme le disait plus haut jpclrec39

amha, pas besoin du .OBJ , du moins pas prioritaire, dans l'ordre STL, DXF, SVG, et sur le code les .INO , GCODE serait top (peut-être moi qui l'avait demandé je sais plus), ce sont des fichiers très utilisés par les #makers fablab

**Comment by jpclerc39** [comment]
> Concernant les aperçus des fichiers .STL :
> Lorsque l'on voit l'image on voit le bouton en bas aperçu interactif et inversement.
> Perso ça me fais bizarre de le voir dans ce sens, on a l'impression que c'est une légende et que l'on est sur l'autre aperçu.
> Peut-on imaginer à la place un bouton double interactif / image avec un côté grisé et un côté coloré en fonction de celui sélectionné... comme ça on voit les deux affichages possibles et celui actif qui correspond à la vue réelle.
> Je viens d'aller sur Tinkercad, voici leur aperçu, la vue affichée est encadrée en dessous en couleur, c'est plus raccord je trouve.
> ![aperçu stl](https://user-images.githubusercontent.com/29262866/78937260-77257e00-7ab0-11ea-98e5-64cce4f3e939.jpg)

+1

**Comment by louis-ev** [comment]
En attendant mieux j’ai ajouté un petit switch qui devrait être + clair.

![image](https://user-images.githubusercontent.com/1948417/79775378-71435e80-8334-11ea-85da-8fd41f8b1ab3.png)

**Comment by louis-ev** [comment]
Finalement après discussion avec @paulinegourlet on a simplifié encore :

<img width="261" alt="image" src="https://user-images.githubusercontent.com/1948417/79779952-7d7eea00-833b-11ea-9f6d-67e37d96b81b.png">

**Comment by jubonhomme** [comment]
Super !

**Comment by jpclerc39** [comment]
whaouh top, je continue à faire joujou avec, et attend impatiemment une petite release à mettre en prod ^_^

**Comment by louis-ev** [comment]
Mais oui !
Il me reste des trucs un peu important avant que je puisse vous envoyer une version, notamment comment créer des admins et gérer les channels de discussion, mais ça arrive dans les jours qui viennent.

**Comment by louis-ev** [comment]
Bon, on a un soucis important : la bibliothèque qui permet de générer des aperçus de fichiers STL (node-stl-to-thumbnail) fait appel à des dépendances qui rendent quasi impossible la création de la version appli de do•doc.
Je vais devoir trouver une autre solution et réécrire cette brique (argh).

**Comment by jubonhomme** [comment]
Argh... dommage c'était génial comme visualiseur interactif de STL, avec en plus les vignettes en aperçu.
Tu as des pistes pour chercher un remplaçant ?

**Comment by louis-ev** [comment]
Oui ! J’ai testé rapido en utilisant le même système que l’encart « STL interactif » avec le moteur qui génère les pdf (électron et puppeteer, qui simulent une fenêtre de navigateur et peuvent faire un screenshot du rendu). Ça fonctionne et ça ajoute pas de dépendance donc idéal.
A tester plus rigoureusement mais a priori on est bon :)

**Comment by louis-ev** [comment]
Implémenté ! ça sera encore mieux quand Electron sera sur la dernière version (sur ma machine virtuelle sous Linux j’ai pas les aperçus car la version d’Electron la gère pas) mais ça fonctionne et je peux exporter à nouveau.

**Comment by jpclerc39** [comment]
génial merci, c'est une fonctionnalité devenue, même avant sa mise en prod, essentielle. on peut retester quelques fichiers stl sur le dodoc de test si besoin

**Comment by louis-ev** [comment]
Chouette :)
Pour la version "serveur" de do•doc pas moyen d’utiliser puppeteer — cela veut dire que je reste sur `node-stl-to-thumbnail`. C’est pas grave mais ça veut dire que le rendu visuel des aperçus sera légèrement différent.
On verra plus tard si on peut avoir mieux, c’est déjà pas mal :)

**Comment by louis-ev** [comment]
Je reviens sur cette issue.
Un éditeur de code pourrait être intéressant, mais il faudrait clarifier le besoin et les fonctionnalités nécessaires (collaboratif ? colorisation du texte ?).
Je la report à dodoc 10

**Comment by jubonhomme** [comment]
Pour les classes de primaires / collèges / lycées, on utilise des versions de type Scratch 
Je sais qu'il existe des plateformes qui reprennent cette interface mais je ne sais pas ce qu'il y a derrière

voir par exemple : 
https://fr.vittascience.com/code
https://makecode.microbit.org/#editor

**Comment by louis-ev** [comment]
À mon avis on déborde largement du cadre de do•doc… non ?

**Comment by jubonhomme** [comment]
Euh oui... je m'explique 
En ce moment tous les projets que je mène et qui utilisent du code avec les écoles se font à partir de plateformes type Microbit et Scratch. Donc effectivement on est complètement hors cadre par rapport à dodoc. 

Pour notre usage l'éditeur de code n'est pas prioritaire. 
Je vois plutôt la complémentarité de dodoc pour la capture de son ou d'image pour un transfert vers une plateforme de type scratch.
Ou la capture d'écran depuis dodoc pour garder une trace du projet sous forme d'image de la suite de blocs.

**Comment by louis-ev** [comment]
Je propose de reporter cette issue à 10.1.

Pour info :
- les STL sont pris en charge
- les OBJ aussi mais sans aperçu image, juste une vue interactive
- DXF ça me semble compliqué, je n'ai pas trouvé d'exemple d'un visualizer web ? pouvez-vous en trouver un ?
- code source type `.ino` : que faire ? Qu'on puisse le modifier comme un texte, avec historique ? Aucune option de mise en forme (type titre ou bold), juste du texte avec un typo mono, comme https://pastebin.com/ ?
